core/node/da_dispatcher/src/da_dispatcher.rs

```diff
@@ -0,0 +1,196 @@
+use std::{future::Future, time::Duration};
+
+use anyhow::Context;
+use chrono::{NaiveDateTime, Utc};
+use rand::Rng;
+use tokio::sync::watch;
+use zksync_config::DADispatcherConfig;
+use zksync_da_client::{
+    types::{DAError, IsTransient},
+    DataAvailabilityClient,
+};
+use zksync_dal::{ConnectionPool, Core, CoreDal};
+use zksync_types::L1BatchNumber;
+
+use crate::metrics::METRICS;
+
+#[derive(Debug)]
+pub struct DataAvailabilityDispatcher {
+    client: Box<dyn DataAvailabilityClient>,
+    pool: ConnectionPool<Core>,
+    config: DADispatcherConfig,
+}
+
+impl DataAvailabilityDispatcher {
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        config: DADispatcherConfig,
+        client: Box<dyn DataAvailabilityClient>,
+    ) -> Self {
+        Self {
+            pool,
+            config,
+            client,
+        }
+    }
+
+    pub async fn run(self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let pool = self.pool.clone();
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("Stop signal received, da_dispatcher is shutting down");
+                break;
+            }
+
+            tokio::join!(
+                async {
+                    if let Err(err) = self.dispatch(&pool).await {
+                        tracing::error!("dispatch error {err:?}");
+                    }
+                },
+                async {
+                    if let Err(err) = self.poll_for_inclusion(&pool).await {
+                        tracing::error!("poll_for_inclusion error {err:?}");
+                    }
+                }
+            );
+
+            tokio::time::sleep(self.config.polling_interval()).await;
```

Comment:
If the polling interval can be reasonably large, it would be better to use `tokio::time::timeout` as in other components, so that the node is more responsive during shutdown:

```rust
if tokio::time::timeout(self.sleep_interval, stop_receiver.changed())
    .await
    .is_ok()
{
    // Stop signal received
}
```